webapp/channels/src/utils/popouts/browser_popouts.ts

```diff
@@ -0,0 +1,168 @@
+                return false;
+            });
+            const closeListener = () => {
+                window.opener.postMessage({
+                    channel: '_close',
+                    args: [],
+                }, window.location.origin);
+            };
+
+            window.addEventListener('beforeunload', closeListener);
+            return () => {
+                unblockHistory();
+                window.removeEventListener('beforeunload', closeListener);
+            };
+        }
+
+        return () => {};
+    }, [history]);
+}
+
+export class BrowserPopouts {
+    private popoutWindows: Map<Window, string>;
+    private listeners: Map<string, Set<(channel: string, ...args: any[]) => void>>;
+    private closeListeners: Map<string, Set<() => void>>;
+
+    constructor() {
+        this.popoutWindows = new Map();
+        this.listeners = new Map();
+        this.closeListeners = new Map();
+
+        window.addEventListener('message', (event) => {
+            const popout = event.source as Window;
+            const target = this.popoutWindows.get(popout);
+
+            if (event.origin !== window.location.origin) {
+                return;
+            }
+
+            if (!target) {
+                return;
+            }
+
+            switch (event.data.channel) {
+            case '_navigate':
+                window.focus();
+                getHistory().push(event.data.args[0]);
+                return;
+            case '_close': {
+                // Unfortunately, there is no way to know if the window is
+                // closed other than checking the beforeunload event
+                // so we need to poll until it is closed or we reach the max tries,
+                // to ensure the close listener is called at the right time and that
+                // the listeners are not deleted prematurely.
+                let tries = 0;
+                const checkInterval = setInterval(() => {
+                    tries++;
+                    if (popout.closed) {
+                        clearInterval(checkInterval);
+                        this.closeListeners.get(target)?.forEach((listener) => {
+                            listener();
+                        });
+                        this.listeners.delete(target);
+                        this.closeListeners.delete(target);
+                        this.popoutWindows.delete(popout);
+                    }
+                    if (tries >= CLOSE_POLL_MAX_TRIES) {
+                        clearInterval(checkInterval);
+                    }
+                }, CLOSE_POLL_INTERVAL);
+                return;
+            }
+            }
+
+            this.listeners.get(target)?.forEach((listener) => {
+                listener(event.data.channel, ...event.data.args);
+            });
+        });
+    }
+
+    setupBrowserPopout = (path: string) => {
+        const target = path.replace(/\//g, '-').replace('-_popout-', '');
```

Comment:
I'm not exactly sure what this replacement is meant to be doing with the two replace calls. If it's supposed to be removing the `/_popout/` from the URL directly, you could do that directly with
```suggestion
        const target = path.replace(/^\/_popout\//g, '');
```